### PR TITLE
op-node: Skip resetting the scheduler timer and channel when a desired action is already pending

### PR DESCRIPTION
**Pull request by ajsutton**
**Description**

Each time the driver goes round it's loop, it asks the sequencer when it wants to perform its next action and then resets the scheduler timer to that delay before it proceeds to a `select` which checks which of the possible channels (including the sequencer timer) is ready for execution.

This introduces a race condition between the sequencer timing firing again after calling `Reset` (which may happen asynchronously even when the delay is 0) and the `select` checking the sequencer timer channel. If the timer doesn't fire in time, the driver goes back round the loop and will reset the sequencer timer again, restarting the race.

To avoid this race, if the sequencer requests a delay of 0 either:
* if the channel already has a pending entry, do nothing - the action is already scheduled correctly
* otherwise create a new channel and synchronously publish to it, guaranteeing it will be ready to execute

**Tests**

Please describe any tests you've added. If you've added no tests, or left important behavior untested, please explain why not.

**Invariants**

For changes to critical code paths, please list and describe the invariants or key security properties of your new or changed code.

**Additional context**

Add any other context about the problem you're solving.

**Metadata**

- Fixes #[Link to Issue]

**TODOs**

- [ ] Author or reviewer has added an entry to the [current release notes draft][RND], if appropriate.

[RND]: https://www.notion.so/oplabs/ded30107ceec41c88817e60322aa8d0a?v=b4a22cedb85a46a38c9be14e7c984953&pvs=4 "Release Notes"

### CODE REVIEW

File: op-node/rollup/driver/state.go [hunk 1]
```diff
@@ -192,8 +192,21 @@ func (s *Driver) eventLoop() {
 	var sequencerCh <-chan time.Time
 	planSequencerAction := func() {
 		delay := s.sequencer.PlanNextSequencerAction()
+		if delay == 0 {
+			if len(sequencerCh) > 0 {
+				// We want to perform the next action immediately and it's already scheduled so just leave it alone.
+				return
+			}
+			// Otherwise create a channel and trigger it immediately.
```
Comment (trianglesphere) [inline op-node/rollup/driver/state.go:200]: Should probably also stop the timer in this case.

File: op-node/rollup/driver/state.go [hunk 2]
```diff
@@ -192,8 +192,21 @@ func (s *Driver) eventLoop() {
...
+			ch := make(chan time.Time, 1)
+			sequencerCh = ch
+			ch <- time.Now()
+			return
+		}
+		// Setup the timer to trigger after the requested delay
 		sequencerCh = sequencerTimer.C
-		if len(sequencerCh) > 0 { // empty if not already drained before resetting
+		if len(sequencerCh) > 0 {
```
[the remark this answers is not in the document]
Comment (ajsutton) [reply op-node/rollup/driver/state.go:208]: Yeah it's probably worth doing. It probably doesn't make a lot of difference in this case - the worst case is that the timer fires between cleaning the channel and resetting the timer which causes us to run a scheduler action sooner than we should. Depends if the sequencer action is smart enough to handle that case or not, but definitely simpler if we just get it right.

File: op-node/rollup/driver/state.go (hunk 1 above)
Comment (ajsutton) [reply op-node/rollup/driver/state.go:200]: Yeah worth while so we don't leave it hanging around only to ignore its result.

**Comment by ajsutton** [comment]
Add do-not-merge label so that we can be deliberate about when to merge this rather than having mergify kick in.

**Comment by trianglesphere** [comment]
IMO this change is probably net positive, but not strictly needed. I think we'll also want to follow up with this later.

@protolambda what do you think about bringing in this change now?

**Comment by protolambda** [comment]
> IMO this change is probably net positive, but not strictly needed. I think we'll also want to follow up with this later.

Agreed

> @protolambda what do you think about bringing in this change now?

I'd avoid making changes to it now. We are free to update the sequencing timing, but should prefer stability so close to launch.

**Comment by ajsutton** [comment]
Agreed, closing for now to avoid PRs hanging around.

**Pull request closed without merging**